File: ws/first_ws/src/ctrl_turtle2_pkg/src/MainWindow.cpp
//
// Created by zcb on 2021/4/10.
//

#include "MainWindow.h"
#include <std_srvs/Empty.h>
#include <turtlesim/Spawn.h>
#include <turtlesim/Kill.h>
#include <turtlesim/SetPen.h>
#include <turtlesim/TeleportAbsolute.h>
#include <turtlesim/TeleportRelative.h>

MainWindow::MainWindow(ros::NodeHandle *node, QWidget *parent) : QWidget(parent) {
    this->setWindowTitle("小乌龟控制");
    this->setFixedSize(500, 300);

    // 关于 QTimer
    this->updateTimer = new QTimer(this);
    this->updateTimer->setInterval(16); // 1000 / 60fps = 16 ms
    QTimer::connect(this->updateTimer, &QTimer::timeout, this, &MainWindow::onUpdate);
    this->updateTimer->start();

    this->node = node;

    // ui 布局
    this->setUI();

    // connect to slot func
    this->connectSlot();

}

MainWindow::~MainWindow() {

}

void MainWindow::onUpdate() {
    // ros 回调
    ros::spinOnce();
    // qt 回调
    this->update();

    if (!ros::ok()) {
        this->close();
    }
}


void MainWindow::setUI() {
    this->ly = new QFormLayout(this);

    clearBtn = new QPushButton("清理画布");
    this->ly->addRow(clearBtn);

    resetBtn = new QPushButton("重置");
    this->ly->addRow(resetBtn);

    QHBoxLayout *h_ly_spawn = new QHBoxLayout();
    spawnX = new QLineEdit();
    spawnX->setPlaceholderText("X坐标");
    spawnY = new QLineEdit();
    spawnY->setPlaceholderText("Y坐标");
    spawnTheta = new QLineEdit();
    spawnTheta->setPlaceholderText("角度");
    spawnName = new QLineEdit();
    spawnName->setPlaceholderText("名字");
    spawnBtn = new QPushButton("创建小乌龟");
    h_ly_spawn->addWidget(spawnX);
    h_ly_spawn->addWidget(spawnY);
    h_ly_spawn->addWidget(spawnTheta);
    h_ly_spawn->addWidget(spawnName);
    h_ly_spawn->addWidget(spawnBtn);
    this->ly->addRow(h_ly_spawn);

    QHBoxLayout *h_ly_kill = new QHBoxLayout();
    killName = new QLineEdit();
    killName->setPlaceholderText("名字");
    killBtn = new QPushButton("杀死小乌龟");
    h_ly_kill->addWidget(killName);
    h_ly_kill->addWidget(killBtn);
    this->ly->addRow(h_ly_kill);

    QHBoxLayout *h_ly_set_pen = new QHBoxLayout();
    setPenName = new QLineEdit();
    setPenName->setPlaceholderText("名字");
    setPenR = new QLineEdit();
    setPenR->setPlaceholderText("红");
    setPenG = new QLineEdit();
    setPenG->setPlaceholderText("绿");
    setPenB = new QLineEdit();
    setPenB->setPlaceholderText("蓝");
    setPenWidth = new QLineEdit();
    setPenWidth->setPlaceholderText("粗细");
    setPenOff = new QCheckBox("关闭");
    setPenBtn = new QPushButton("设置画笔");
    h_ly_set_pen->addWidget(setPenName);
    h_ly_set_pen->addWidget(setPenR);
    h_ly_set_pen->addWidget(setPenG);
    h_ly_set_pen->addWidget(setPenB);
    h_ly_set_pen->addWidget(setPenWidth);
    h_ly_set_pen->addWidget(setPenOff);
    h_ly_set_pen->addWidget(setPenBtn);
    this->ly->addRow(h_ly_set_pen);


    QHBoxLayout *h_ly_abs = new QHBoxLayout();
    abs_name = new QLineEdit();
    abs_name->setPlaceholderText("名字");
    abs_x = new QLineEdit();
    abs_x->setPlaceholderText("X坐标");
    abs_y = new QLineEdit();
    abs_y->setPlaceholderText("Y坐标");
    abs_theta = new QLineEdit();
    abs_theta->setPlaceholderText("角度");
    teleportAbsoluteBtn = new QPushButton("设置绝对位置");
    h_ly_abs->addWidget(abs_name);
    h_ly_abs->addWidget(abs_x);
    h_ly_abs->addWidget(abs_y);
    h_ly_abs->addWidget(abs_theta);
    h_ly_abs->addWidget(teleportAbsoluteBtn);
    this->ly->addRow(h_ly_abs);


    QHBoxLayout *h_ly_rel = new QHBoxLayout();
    rel_name = new QLineEdit();
    rel_name->setPlaceholderText("名字");
    rel_linear = new QLineEdit();
    rel_linear->setPlaceholderText("线速度");
    rel_angular = new QLineEdit();
    rel_angular->setPlaceholderText("角速度");
    teleportRelativeBtn = new QPushButton("设置相对位置");
    h_ly_rel->addWidget(rel_name);
    h_ly_rel->addWidget(rel_linear);
    h_ly_rel->addWidget(rel_angular);
    h_ly_rel->addWidget(teleportRelativeBtn);
    this->ly->addRow(h_ly_rel);

}


void MainWindow::connectSlot() {
    QPushButton::connect(this->clearBtn, &QPushButton::clicked, this, &MainWindow::clickClear);
    QPushButton::connect(this->resetBtn, &QPushButton::clicked, this, &MainWindow::clickReset);
    QPushButton::connect(this->spawnBtn, &QPushButton::clicked, this, &MainWindow::clickSpawn);
    QPushButton::connect(this->killBtn, &QPushButton::clicked, this, &MainWindow::clickKill);
    QPushButton::connect(this->setPenBtn, &QPushButton::clicked, this, &MainWindow::clickSetPen);
    QPushButton::connect(this->teleportAbsoluteBtn, &QPushButton::clicked, this, &MainWindow::clickAbs);
    QPushButton::connect(this->teleportRelativeBtn, &QPushButton::clicked, this, &MainWindow::clickRel);
}

void MainWindow::clickClear() {

    ros::ServiceClient clear_client = node->serviceClient<std_srvs::Empty>("/clear");
    std_srvs::Empty service;
    clear_client.call(service);

    clear_client.shutdown(); // 关闭这个临时客户端
}

void MainWindow::clickReset() {

    ros::ServiceClient reset_client = node->serviceClient<std_srvs::Empty>("/reset");
    std_srvs::Empty service;
    reset_client.call(service);

    reset_client.shutdown();
}

void MainWindow::clickSpawn() {
    ros::ServiceClient spawn_client = node->serviceClient<turtlesim::Spawn>("/spawn");
    turtlesim::Spawn service;
    service.request.x = this->spawnX->text().toFloat();
    service.request.y = this->spawnY->text().toFloat();
    service.request.theta = this->spawnTheta->text().toFloat() * M_PI / 180;
    service.request.name = this->spawnName->text().toStdString();

    spawn_client.call(service);

    spawn_client.shutdown();
}

void MainWindow::clickKill() {
    ros::ServiceClient kill_client = node->serviceClient<turtlesim::Kill>("/kill");
    turtlesim::Kill service;
    service.request.name = this->killName->text().toStdString();

    kill_client.call(service);

    kill_client.shutdown();
}

void MainWindow::clickSetPen() {
    std::string nodeName = this->setPenName->text().toStdString();
    ros::ServiceClient set_pen_client = node->serviceClient<turtlesim::SetPen>("/" + nodeName + "/set_pen");
    turtlesim::SetPen service;
    service.request.r = this->setPenR->text().toUInt();
    service.request.g = this->setPenG->text().toUInt();
    service.request.b = this->setPenB->text().toUInt();
    service.request.width = this->setPenWidth->text().toUInt();
    service.request.off = this->setPenOff->isChecked();

    set_pen_client.call(service);
    set_pen_client.shutdown();
}

void MainWindow::clickAbs() {
    std::string nodeName = this->abs_name->text().toStdString();
    ros::ServiceClient abs_client = node->serviceClient<turtlesim::TeleportAbsolute>(
            "/" + nodeName + "/teleport_absolute");
    turtlesim::TeleportAbsolute service;
    service.request.x = this->abs_x->text().toFloat();
    service.request.y = this->abs_y->text().toFloat();
    service.request.theta = this->abs_theta->text().toFloat() * M_PI / 180;

    abs_client.call(service);
    abs_client.shutdown();
}

void MainWindow::clickRel() {
    std::string nodeName = this->rel_name->text().toStdString();
    ros::ServiceClient rel_client = node->serviceClient<turtlesim::TeleportRelative>(
            "/" + nodeName + "/teleport_relative");
    turtlesim::TeleportRelative service;
    service.request.linear = this->rel_linear->text().toFloat();
    service.request.angular = this->rel_angular->text().toFloat() * M_PI / 180;
    rel_client.call(service);

    rel_client.shutdown();
}

File: ws/first_ws/src/ctrl_turtle2_pkg/scripts/MainWindow.py
#! /usr/bin/env python
# coding:utf8
from PyQt5.QtWidgets import (
    QWidget, QFormLayout, QPushButton,
    QCheckBox, QLineEdit, QHBoxLayout,
)
from PyQt5.QtCore import (
    QTimer
)
import rospy
# srv
from std_srvs.srv import Empty, EmptyRequest
from turtlesim.srv import (
    Spawn, SpawnRequest, Kill, KillRequest,
    SetPen, SetPenRequest,
    TeleportAbsolute, TeleportAbsoluteRequest,
    TeleportRelative, TeleportRelativeRequest,
)


class MainWindow(QWidget):
    def __init__(self):
        super(MainWindow, self).__init__()
        self.setWindowTitle("小乌龟控制")
        self.setFixedSize(500, 300)

        # QTimer
        self.timer = QTimer()
        self.timer.setInterval(16)  # 1000 /60fps = 16
        self.timer.timeout.connect(self.onUpdate)
        self.timer.start()

        # ui
        self.setUI()

        # connect to slot
        self.connectSlot()

    def setUI(self):
        self.ly = QFormLayout(self)

        self.clearBtn = QPushButton("清理画布")
        self.ly.addRow(self.clearBtn)

        self.resetBtn = QPushButton("重置")
        self.ly.addRow(self.resetBtn)

        h_ly_spawn = QHBoxLayout()
        self.spawnX = QLineEdit()
        self.spawnX.setPlaceholderText("X坐标")
        self.spawnY = QLineEdit()
        self.spawnY.setPlaceholderText("Y坐标")
        self.spawnTheta = QLineEdit()
        self.spawnTheta.setPlaceholderText("角度")
        self.spawnName = QLineEdit()
        self.spawnName.setPlaceholderText("名字")
        self.spawnBtn = QPushButton("创建小乌龟")
        h_ly_spawn.addWidget(self.spawnX)
        h_ly_spawn.addWidget(self.spawnY)
        h_ly_spawn.addWidget(self.spawnTheta)
        h_ly_spawn.addWidget(self.spawnName)
        h_ly_spawn.addWidget(self.spawnBtn)
        self.ly.addRow(h_ly_spawn)

        h_ly_kill = QHBoxLayout()
        self.killName = QLineEdit()
        self.killName.setPlaceholderText("名字")
        self.killBtn = QPushButton("杀死小乌龟")
        h_ly_kill.addWidget(self.killName)
        h_ly_kill.addWidget(self.killBtn)
        self.ly.addRow(h_ly_kill)

        h_ly_set_pen = QHBoxLayout()
        self.setPenName = QLineEdit()
        self.setPenName.setPlaceholderText("名字")
        self.setPenR = QLineEdit()
        self.setPenR.setPlaceholderText("红")
        self.setPenG = QLineEdit()
        self.setPenG.setPlaceholderText("绿")
        self.setPenB = QLineEdit()
        self.setPenB.setPlaceholderText("蓝")
        self.setPenWidth = QLineEdit()
        self.setPenWidth.setPlaceholderText("粗细")
        self.setPenOff = QCheckBox("关闭")
        self.setPenBtn = QPushButton("设置画笔")
        h_ly_set_pen.addWidget(self.setPenName)
        h_ly_set_pen.addWidget(self.setPenR)
        h_ly_set_pen.addWidget(self.setPenG)
        h_ly_set_pen.addWidget(self.setPenB)
        h_ly_set_pen.addWidget(self.setPenWidth)
        h_ly_set_pen.addWidget(self.setPenOff)
        h_ly_set_pen.addWidget(self.setPenBtn)
        self.ly.addRow(h_ly_set_pen)

        h_ly_abs = QHBoxLayout()
        self.abs_name = QLineEdit()
        self.abs_name.setPlaceholderText("名字")
        self.abs_x = QLineEdit()
        self.abs_x.setPlaceholderText("X坐标")
        self.abs_y = QLineEdit()
        self.abs_y.setPlaceholderText("Y坐标")
        self.abs_theta = QLineEdit()
        self.abs_theta.setPlaceholderText("角度")
        self.teleportAbsoluteBtn = QPushButton("设置绝对位置")
        h_ly_abs.addWidget(self.abs_name)
        h_ly_abs.addWidget(self.abs_x)
        h_ly_abs.addWidget(self.abs_y)
        h_ly_abs.addWidget(self.abs_theta)
        h_ly_abs.addWidget(self.teleportAbsoluteBtn)
        self.ly.addRow(h_ly_abs)

        h_ly_rel = QHBoxLayout()
        self.rel_name = QLineEdit()
        self.rel_name.setPlaceholderText("名字")
        self.rel_linear = QLineEdit()
        self.rel_linear.setPlaceholderText("线速度")
        self.rel_angular = QLineEdit()
        self.rel_angular.setPlaceholderText("角速度")
        self.teleportRelativeBtn = QPushButton("设置相对位置")
        h_ly_rel.addWidget(self.rel_name)
        h_ly_rel.addWidget(self.rel_linear)
        h_ly_rel.addWidget(self.rel_angular)
        h_ly_rel.addWidget(self.teleportRelativeBtn)
        self.ly.addRow(h_ly_rel)

    def onUpdate(self):
        # qt 回调
        self.update()

        if rospy.is_shutdown():
            self.close()

    def connectSlot(self):
        self.clearBtn.clicked.connect(self.clickClear)
        self.resetBtn.clicked.connect(self.clickReset)
        self.spawnBtn.clicked.connect(self.clickSpawn)
        self.killBtn.clicked.connect(self.clickKill)
        self.setPenBtn.clicked.connect(self.clickSetPen)
        self.teleportAbsoluteBtn.clicked.connect(self.clickAbs)
        self.teleportRelativeBtn.clicked.connect(self.clickRel)

    def clickClear(self):
        clear_client = rospy.ServiceProxy("/clear", Empty)
        request = EmptyRequest()
        clear_client.call(request)
        clear_client.close()  # 关闭临时的 client

    def clickReset(self):
        reset_client = rospy.ServiceProxy("/reset", Empty)
        request = EmptyRequest()
        reset_client.call(request)
        reset_client.close()

    def clickSpawn(self):
        spawn_client = rospy.ServiceProxy("/spawn", Spawn)
        request = SpawnRequest()

        request.x = float(self.spawnX.text())
        request.y = float(self.spawnY.text())
        request.theta = float(self.spawnTheta.text()) * 3.14 / 180
        request.name = self.spawnName.text()

        spawn_client.call(request)
        spawn_client.close()

    def clickKill(self):
        kill_client = rospy.ServiceProxy("/kill", Kill)
        request = KillRequest()

        request.name = self.killName.text()

        kill_client.call(request)
        kill_client.close()

    def clickSetPen(self):
        set_pen_client = rospy.ServiceProxy("/" + self.setPenName.text() +
                                            "/set_pen", SetPen)
        request = SetPenRequest()

        request.r = float(self.setPenR.text())
        request.g = float(self.setPenG.text())
        request.b = float(self.setPenB.text())
        request.width = float(self.setPenWidth.text())
        request.off = float(self.setPenOff.isChecked())

        set_pen_client.call(request)
        set_pen_client.close()

    def clickAbs(self):
        abs_client = rospy.ServiceProxy("/" + self.abs_name.text() +
                                        "/teleport_absolute", TeleportAbsolute)
        request = TeleportAbsoluteRequest()

        request.x = float(self.abs_x.text())
        request.y = float(self.abs_y.text())
        request.theta = float(self.abs_theta.text()) * 3.14 / 180

        abs_client.call(request)
        abs_client.close()

    def clickRel(self):
        rel_client = rospy.ServiceProxy("/" + self.rel_name.text() +
                                        "/teleport_relative", TeleportRelative)
        request = TeleportRelativeRequest()
        request.linear = float(self.rel_linear.text())
        request.angular = float(self.rel_angular.text()) * 3.14 / 180

        rel_client.call(request)
        rel_client.close()
